ulm-semantics/test/execution.md

```diff
@@ -63,68 +69,71 @@ module ULM-TEST-EXECUTION
-    rule <k> hold_list_values_from_test_stack => list_ptrs_holder L ~> list_values_holder .List ... </k>
-         <test-stack> L:List => .List </test-stack>
-    rule <k> list_ptrs_holder ListItem(I) LPH ~> list_values_holder LLH
-                => I ~> list_ptrs_holder LPH ~> list_values_holder LLH ... </k>
-    rule <k> ptrValue(_, V) ~> list_ptrs_holder LPH ~> list_values_holder LLH
-                => list_ptrs_holder LPH ~> list_values_holder ListItem(V) LLH ... </k>
-    rule <k> list_ptrs_holder .List => .K ... </k>
-
-    rule <k> hold I => value_holder I ... </k>
-
-    rule <k> encode_call_data_to_string
-             ~> list_values_holder ARGS , list_values_holder PTYPES , value_holder FNAME , .ULMTestTypeHolderList
-             => Bytes2String(encodeCallData(FNAME, PTYPES, ARGS))
-            ...
-         </k>
-
-    rule <k> encode_call_data_to_string
-             ~> value_holder FNAME
-             => Bytes2String(encodeCallData(FNAME, .List, .List))
-            ...
-         </k> [owise]
+    syntax BytesOrError ::= extractCallSignature(EncodeCall)  [function, total]
+    rule extractCallSignature(Fn:Identifier ( Args:EncodeArgs ))
+        => methodSignature(IdentifierToString(Fn), encodeArgsToNormalizedParams(Args))
+    syntax NormalizedFunctionParameterList ::= encodeArgsToNormalizedParams(EncodeArgs)  [function, total]
+    rule encodeArgsToNormalizedParams(.EncodeArgs) => .NormalizedFunctionParameterList
+    rule encodeArgsToNormalizedParams(_:Expression : T:Type ,  Eas:EncodeArgs )
+        => #token("#unused", "Identifier"):T , encodeArgsToNormalizedParams(Eas)
+
+    syntax Identifier ::= "append_bytes_raw"  [token]
+                        | "buffer_id"  [token]
+                        | "bytes_hooks"  [token]
+                        | "empty"  [token]
+                        | "str"  [token]
+
+    syntax EncodeValues ::= encodeArgsToEncodeValues(EncodeArgs)  [function, total]
+    rule encodeArgsToEncodeValues(.EncodeArgs) => .EncodeValues
+    rule encodeArgsToEncodeValues(E:Expression : T:Type , Es:EncodeArgs)
+        => E : T , encodeArgsToEncodeValues(Es)
+
+    syntax NonEmptyStatementsOrError ::= encodeInstructions(EncodeCall)  [function, total]
+    rule encodeInstructions( _:Identifier ( Args:EncodeArgs ) )
+        => encodeInstructions(Args)
+    syntax NonEmptyStatementsOrError ::= encodeInstructions(EncodeArgs)  [function, total]
+    rule encodeInstructions(Args:EncodeArgs)
+        => concat
+            (   let buffer_id = :: bytes_hooks :: empty( .CallParamsList );
+                .NonEmptyStatements
+            ,   codegenValuesEncoder(buffer_id, encodeArgsToEncodeValues(Args))
+            )
+
+    rule encode_call_data C:EncodeCall
+        => encodeCallData(extractCallSignature(C), encodeInstructions(C))
+
+    syntax ExecutionItem ::= encodeCallData(BytesOrError, NonEmptyStatementsOrError)
```

Comment:
It can't have a "total" attribute because it's not defined when either the first or the second argument is an error. Since partial functions have all sorts of problems (e.g. having a `Left => Right` rewrite rule in a semantics where `Right` may be undefined makes the entire semantics unsound, or whatever is the proper word for it), I would prefer to use them only when absolutely needed.

Even if we ignore that, I would argue that not making this a function makes the code more debugable. When evaluating an undefined function, the LLVM backend will print the function that it can't evaluate, which is sort of nice. However, it's harder to get the entire context in which this happened (i.e., the entire configuration). If `encodeCallData` is not a function, you get the entire configuration.

With the Haskell backend, if we make this a function, it depends. If there is a `#Ceil` definition for this, then proving would just succeed (since bottom implies everything), although that's certainly not what we want here (although I think that they may have added a warning for this case, or something like that). Without a `#Ceil` definition, you'd get the full configuration with an unevaluated function, which is fine for debugging.